File: en/codes/csharp/chapter_tree/array_binary_tree.cs
/**
* File: array_binary_tree.cs
* Created Time: 2023-07-20
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_tree;

/* Binary tree class represented by array */
public class ArrayBinaryTree(List<int?> arr) {
    List<int?> tree = new(arr);

    /* List capacity */
    public int Size() {
        return tree.Count;
    }

    /* Get value of node at index i */
    public int? Val(int i) {
        // If index out of bounds, return null to represent empty position
        if (i < 0 || i >= Size())
            return null;
        return tree[i];
    }

    /* Get index of left child node of node at index i */
    public int Left(int i) {
        return 2 * i + 1;
    }

    /* Get index of right child node of node at index i */
    public int Right(int i) {
        return 2 * i + 2;
    }

    /* Get index of parent node of node at index i */
    public int Parent(int i) {
        return (i - 1) / 2;
    }

    /* Level-order traversal */
    public List<int> LevelOrder() {
        List<int> res = [];
        // Traverse array directly
        for (int i = 0; i < Size(); i++) {
            if (Val(i).HasValue)
                res.Add(Val(i)!.Value);
        }
        return res;
    }

    /* Depth-first traversal */
    void DFS(int i, string order, List<int> res) {
        // If empty position, return
        if (!Val(i).HasValue)
            return;
        // Preorder traversal
        if (order == "pre")
            res.Add(Val(i)!.Value);
        DFS(Left(i), order, res);
        // Inorder traversal
        if (order == "in")
            res.Add(Val(i)!.Value);
        DFS(Right(i), order, res);
        // Postorder traversal
        if (order == "post")
            res.Add(Val(i)!.Value);
    }

    /* Preorder traversal */
    public List<int> PreOrder() {
        List<int> res = [];
        DFS(0, "pre", res);
        return res;
    }

    /* Inorder traversal */
    public List<int> InOrder() {
        List<int> res = [];
        DFS(0, "in", res);
        return res;
    }

    /* Postorder traversal */
    public List<int> PostOrder() {
        List<int> res = [];
        DFS(0, "post", res);
        return res;
    }
}

public class array_binary_tree {
    [Test]
    public void Test() {
        // Initialize binary tree
        // Here we use a function to generate a binary tree directly from an array
        List<int?> arr = [1, 2, 3, 4, null, 6, 7, 8, 9, null, null, 12, null, null, 15];

        TreeNode? root = TreeNode.ListToTree(arr);
        Console.WriteLine("\nInitialize binary tree\n");
        Console.WriteLine("Array representation of binary tree:");
        Console.WriteLine(arr.PrintList());
        Console.WriteLine("Linked list representation of binary tree:");
        PrintUtil.PrintTree(root);

        // Binary tree class represented by array
        ArrayBinaryTree abt = new(arr);

        // Access node
        int i = 1;
        int l = abt.Left(i);
        int r = abt.Right(i);
        int p = abt.Parent(i);
        Console.WriteLine("\nCurrent node index is " + i + ", value is " + abt.Val(i));
        Console.WriteLine("Its left child node index is " + l + ", value is " + (abt.Val(l).HasValue ? abt.Val(l) : "null"));
        Console.WriteLine("Its right child node index is " + r + ", value is " + (abt.Val(r).HasValue ? abt.Val(r) : "null"));
        Console.WriteLine("Its parent node index is " + p + ", value is " + (abt.Val(p).HasValue ? abt.Val(p) : "null"));

        // Traverse tree
        List<int> res = abt.LevelOrder();
        Console.WriteLine("\nLevel-order traversal is:" + res.PrintList());
        res = abt.PreOrder();
        Console.WriteLine("Preorder traversal is:" + res.PrintList());
        res = abt.InOrder();
        Console.WriteLine("Inorder traversal is:" + res.PrintList());
        res = abt.PostOrder();
        Console.WriteLine("Postorder traversal is:" + res.PrintList());
    }
}

File: ja/codes/java/chapter_tree/array_binary_tree.java
/**
 * File: array_binary_tree.java
 * Created Time: 2023-07-19
 * Author: krahets (krahets@163.com)
 */

package chapter_tree;

import utils.*;
import java.util.*;

/* 配列表現による二分木クラス */
class ArrayBinaryTree {
    private List<Integer> tree;

    /* コンストラクタ */
    public ArrayBinaryTree(List<Integer> arr) {
        tree = new ArrayList<>(arr);
    }

    /* リスト容量 */
    public int size() {
        return tree.size();
    }

    /* インデックス i のノードの値を取得 */
    public Integer val(int i) {
        // インデックスが範囲外なら、空きを表す null を返す
        if (i < 0 || i >= size())
            return null;
        return tree.get(i);
    }

    /* インデックス i のノードの左子ノードのインデックスを取得 */
    public Integer left(int i) {
        return 2 * i + 1;
    }

    /* インデックス i のノードの右子ノードのインデックスを取得 */
    public Integer right(int i) {
        return 2 * i + 2;
    }

    /* インデックス i のノードの親ノードのインデックスを取得 */
    public Integer parent(int i) {
        return (i - 1) / 2;
    }

    /* レベル順走査 */
    public List<Integer> levelOrder() {
        List<Integer> res = new ArrayList<>();
        // 配列を直接走査する
        for (int i = 0; i < size(); i++) {
            if (val(i) != null)
                res.add(val(i));
        }
        return res;
    }

    /* 深さ優先探索 */
    private void dfs(Integer i, String order, List<Integer> res) {
        // 空きスロットなら返す
        if (val(i) == null)
            return;
        // 先行順走査
        if ("pre".equals(order))
            res.add(val(i));
        dfs(left(i), order, res);
        // 中順走査
        if ("in".equals(order))
            res.add(val(i));
        dfs(right(i), order, res);
        // 後順走査
        if ("post".equals(order))
            res.add(val(i));
    }

    /* 先行順走査 */
    public List<Integer> preOrder() {
        List<Integer> res = new ArrayList<>();
        dfs(0, "pre", res);
        return res;
    }

    /* 中順走査 */
    public List<Integer> inOrder() {
        List<Integer> res = new ArrayList<>();
        dfs(0, "in", res);
        return res;
    }

    /* 後順走査 */
    public List<Integer> postOrder() {
        List<Integer> res = new ArrayList<>();
        dfs(0, "post", res);
        return res;
    }
}

public class array_binary_tree {
    public static void main(String[] args) {
        // 二分木を初期化
        // ここでは、配列から直接二分木を生成する関数を利用する
        List<Integer> arr = Arrays.asList(1, 2, 3, 4, null, 6, 7, 8, 9, null, null, 12, null, null, 15);

        TreeNode root = TreeNode.listToTree(arr);
        System.out.println("\n二分木を初期化\n");
        System.out.println("二分木の配列表現:");
        System.out.println(arr);
        System.out.println("二分木の連結リスト表現:");
        PrintUtil.printTree(root);

        // 配列表現による二分木クラス
        ArrayBinaryTree abt = new ArrayBinaryTree(arr);

        // ノードにアクセス
        int i = 1;
        Integer l = abt.left(i);
        Integer r = abt.right(i);
        Integer p = abt.parent(i);
        System.out.println("\n現在のノードのインデックスは " + i + "、値は " + abt.val(i));
        System.out.println("その左子ノードのインデックスは " + l + "、値は " + (l == null ? "null" : abt.val(l)));
        System.out.println("その右子ノードのインデックスは " + r + "、値は " + (r == null ? "null" : abt.val(r)));
        System.out.println("その親ノードのインデックスは " + p + "、値は " + (p == null ? "null" : abt.val(p)));

        // 木を走査
        List<Integer> res = abt.levelOrder();
        System.out.println("\nレベル順走査: " + res);
        res = abt.preOrder();
        System.out.println("前順走査: " + res);
        res = abt.inOrder();
        System.out.println("中順走査: " + res);
        res = abt.postOrder();
        System.out.println("後順走査: " + res);
    }
}
